tests/functional/test_main.py

```diff
@@ -0,0 +1,144 @@
+from cgi import test
+from http import HTTPStatus
+from flask_login import login_user, current_user
+from flask import session
+
+
+def test_root(client, app, test_db):
+
+    response = client.get("/")
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_login_get(client, app):
+
+    response = client.get("login")
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_login_post(client, app, persisted_user):
+
+    response = client.post(
+        "login",
+        data={"username": persisted_user.username, "password": "unhashed-pw"},
+        follow_redirects=True,
+    )
+    assert b"Welcome back" in response.data
+
+
+def test_login_post_bad_pw(client, app, persisted_user):
+
+    response = client.post(
+        "login",
+        data={"username": persisted_user.username, "password": "wrong_password"},
+        follow_redirects=True,
+    )
+    assert b"Invalid credentials" in response.data
+
+
+def test_login_post_bad_username(client, app, persisted_user):
+
+    response = client.post(
+        "login",
+        data={"username": "badusername", "password": "unhashed-pw"},
+        follow_redirects=True,
+    )
+    assert b"Username not found" in response.data
+
+
+def test_signup_get(app):
+
+    with app.test_client() as test_client:
+        response = test_client.get("/signup")
+        assert response.status_code == HTTPStatus.OK
+        assert b"Sign up for Check Out This Song!" in response.data
+
+
+def test_signup_post(client, app, test_db):
+
+    response = client.post(
+        "/signup",
+        data={
+            "username": "testSignupUser",
+            "password": "test_signup_password",
+            "email": "test_email@email.com",
+            "avatar": None,
+            "bio": "",
+        },
+        follow_redirects=True,
+    )
+    print(response.data)
+    assert b"Welcome to Check Out This Song" in response.data
+
+
+def test_signup_post_illegal_username(client, app, test_db):
+
+    response = client.post(
+        "/signup",
+        data={
+            "username": "test_signup_user",
+            "password": "test_signup_password",
+            "email": "test_email@email.com",
+            "avatar": None,
+            "bio": "",
+        },
+        follow_redirects=True,
+    )
+    assert b'<p class="help is-danger">' in response.data
+
+
+def test_user_detail_page(client, app, test_db, persisted_user):
+
+    response = client.get(f"/users/{persisted_user.username}")
+    assert b"test_user1" in response.data
+    assert response.status_code == HTTPStatus.OK
+
+
+def test_user_detail_page_404(client, app):
+
+    response = client.get(f"/users/not_a_real_user")
+    assert b"404" in response.data
+
+
+def test_user_search(client, app):
+    response = client.get("/usersearch")
+    assert b"Search for a user" in response.data
+
+
+def test_user_search_q(client, app, persisted_user):
```

Comment:
Could be good to use a more verbose unit test naming convention

https://dzone.com/articles/7-popular-unit-test-naming

Like test_thing_should_behavior_given_condition

Makes your tests a kind of documentation